minimal_scheduling/minimal_sched_fifo_main_thread.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pthread.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+/* For this example, we will be configuring the main thread scheduling policy with a SCHED_FIFO
+ * real-time scheduling and with real-time priority. The thread middleware threads will inherit
+ * these settings because they are created after the scheduling configuration. */
+
+using namespace std::chrono_literals;
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
+        this->publisher_->publish(message);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+int main(int argc, char * argv[])
+{
+  // The middleware threads are created on the node construction.
+  // Since the main thread settings are configured AFTER the middleware threads
```

Comment:
The before vs after confused me a bit. I got it only after comparing it side by side a few times.

Is it possible to make it more explicit using the [`pthread_attr_setinheritsched`](https://linux.die.net/man/3/pthread_attr_setinheritsched) call with either `PTHREAD_EXPLICIT_SCHED` or `PTHREAD_INHERIT_SCHED`? I took a brief look but I can't find any way unless you create the pthread manually (which is what I do), which may be no good for ROS... If this is not possible, we can leave it as is, as I think your comment is basically as clear as they can be..

Reply:
One option would be to change the main thread attributes to `PTHREAD_EXPLICIT_SCHED` instead to change the configuration order. If that's possible do you think it would be more clear?

Reply:
I think what you have here is fine. In either case the background threads are created by ROS and you can't explicitly pass a scheduler parameter to that?

Reply:
That's right, it's not possible.